### DISCUSSION TITLE
Birdeye view

### DISCUSSION THREAD

**Discussion by ovaisa** [iTwin/itwinjs-core#2231] [Q&A]
I am interested in showing a 'Birdeye view' window in our solution similar to Google Maps (Street view).  We have 3 different use cases, below is a mock up of one of those uses cases.  This use case is same as Cross-Probing example except we would like to use a Birdeye view window rather than split screen.  Other use cases will show a site layout in the birdeye window and use can select a room or building on the site map to zoom to that area in 3D.

Our thought was that we would include the content of birdeye window as a submodel in the iModel (same as is in the Cross Probing example - GeometricModel2d).  We would then display this model in birdeye window and hopefully we can use iTwin capabilities to navigate model (only zoom and pan required) and pick objects.

![Bayview](https://user-images.githubusercontent.com/83007711/133101659-f05f9bde-bf7c-48b0-955c-5d0db7bfb55b.JPG)

**Comment by pmconne** [comment]
Hi @ovaisa. This should all be doable, but you haven't actually asked a question. If the question is "how do I do this", please be specific.

**Reply by ovaisa** [reply]
Can you please guide how I can do this

**Reply by pmconne** [reply]
Yes, if you can be more specific about where you're having trouble.
- A viewport can occupy any div as long as its width and height are greater than zero. If you want it to float, use CSS to float it.
- Is something missing from the cross-probing sample?
- <Your specific question(s) here>

If you've written code and it's not doing what you expect, please post the code and describe what's wrong.

**Reply by ovaisa** [reply]
For the birdeye window, I feel I need two components
- a component to render an imodel, with zoom, pan and element selection capability
- a window to host this component

I created an app from react template to try to implement birdeye using iTwin components.  For the window, I used the Dialog component and for rendering imodel I used the Viewer component in the dialog.  Please see my code below.  My test app is attached at bottom of this post, the code is in BirdEye class in MyFirstUiProvider.tsx

  public static openDialog = () => {
    const iModelId = process.env.IMJS_IMODEL_ID;
    const contextId = process.env.IMJS_CONTEXT_ID;

    ModalDialogManager.openDialog(
      <Dialog
        id={"birdEyeDialog"} title={"BirdEye"}
        opened={true} modal={false} movable={true}
        onClose={BirdEye.closeDialog} onEscape={BirdEye.closeDialog}
        width={400} height={300} alignment={DialogAlignment.Center}>
      <Viewer
          contextId={contextId} iModelId={iModelId}
          authConfig={{ config: BirdEye.authConfig }}
          theme={"light"}
          defaultUiConfig={appUiConfig}
        />
      </Dialog>);
  
  };

I added a widget in the test app with a Button to open this 'birdeye' Dialog.  The dialog opens and shows the 'Your iModel is loading' message.  After a few seconds, a second dialog open up (it is behind the existing dialog).  This new dialog displays the imodel but the Viewer component in the dialog seems to be the one defined in App.tsx (and not the one defined in Dialog which doesn't have any uiProviders).  The original dialog continues to display the loading message and then another dialog opens up.  This keeps repeating (I have to close the dialog to stop).  Pls see image below.

It seems that Dialog and Viewer may not be the right components for this.  Can you please guide me what components I might be able to use for birdeye window and rendering/navigating the iModel.  Perhaps this is not the right approach, in that case would appreciate guidance.

To answer your question on cross probing sample, we feel that a small moveable birdeye window will give more screen space for the 3D model and would work better in our app, especially for other use cases we have planned.

![Birdeye](https://user-images.githubusercontent.com/83007711/133129599-c02f92ce-e1c3-412d-92ce-ecbbfed1b5d7.JPG)

[xviewer.zip](https://github.com/imodeljs/imodeljs/files/7156395/xviewer.zip)

**Reply by ovaisa** [reply]
I was able to make progress by creating a new Screen Viewport with a new ViewState (2D) and adding it to the View Manager.  I am using Dialog to host this viewport.  The P&ID shows up in the viewport and I can select elements in the birdeye window. Some of the navigation toolbars buttons are getting mixed up between the two viewports but generally seems to be working as expected.

Would appreciate feedback if I am on the right track to implement this capability.  Can iTwin team see any potential problems in using these two viewport in this way which might cause some issues later or if there is another better way.

![Birdeye](https://user-images.githubusercontent.com/83007711/133236423-566a15be-4774-4839-97ed-2293f19ff040.JPG)

**Reply by pmconne** [reply]
Thanks for the update. ViewManager is designed to permit any number of viewports. Don't expect pitfalls there (but do report anything that doesn't work as you expect).
I don't use the UI framework or React so I can't comment on potential pitfalls there. @DanEastBentley or @bsteinbk?

**Reply by DanEastBentley** [reply]
You can host a ViewportComponent React component from ui-components in a Dialog, Widget or content group/layout. There are examples in ui-test-app in 
https://github.com/imodeljs/imodeljs/blob/master/test-apps/ui-test-app/src/frontend/appui/dialogs/ViewportDialog.tsx
https://github.com/imodeljs/imodeljs/blob/master/test-apps/ui-test-app/src/frontend/appui/widgets/ViewportWidget.tsx.

**Reply by bsteinbk** [reply]
The Navigation tools and associated Navigation aid monitors the "active" content control in the stage. We are just starting to look at providing a way to designate a "floating" window as an official content control. Currently content controls are expected to be defined in a ContentGroup and displayed within the content area as I show below.
![multi-view-plant](https://user-images.githubusercontent.com/65047615/133258236-6f4e84d1-75ab-4667-898d-a180748d12a7.png)

**Reply by paprophet** [reply]
> 
" We are just starting to look at providing a way to designate a "floating" window as an official content control."
Has  this official content control  been implemented??

**Reply by bsteinbk** [reply]
We still don't have a UX design for a "floating" window, so no official work has been done. Since you already have a floating window implemented, you can get most, if not all, the viewing tools to work by calling IModelApp.viewManager.setSelectedView(floatingViewport); when your floating view has focus. Other tools may use the active content control to retrieve the active viewport so those tools would require official support of a floating content control before they would work properly. @NancyMcCallB @FlyersPh9

**Reply by JohnMajerle** [reply]
From our customer feedback what they would prefer is being to open multiple viewports just like they do in MicroStation for the same reasons that MicroStation lets you do this.  So hardcoding the display to 1, 2 or even 3 up front is not ideal.  One idea I had was to implement a button widget that would let you create new tabs that contained a viewport and a ComboBox model selector. (Unfortunately I didn't have time to prototype it.)

**Reply by NancyMcCallB** [reply]
@JohnMajerle we have a couple of things in our test app that might help you get started. splitSingleViewportCommandDef() in [ToolSpecifications.tsx](https://github.com/iTwin/itwinjs-core/blob/master/test-apps/ui-test-app/src/frontend/tools/ToolSpecifications.tsx) shows how to dynamically change the content layout. The [ViewSelectorPanel](https://github.com/iTwin/itwinjs-core/blob/master/test-apps/ui-test-app/src/frontend/tools/ViewSelectorPanel.tsx) puts the available views in a panel, but you could reuse the code to harvest them for a combo box. I'm not sure what you mean by creating tabs that host a viewport -- do you mean putting the viewport in a widget?

**Reply by JohnMajerle** [reply]
Hey Nancy,

I probably didn't describe that idea very well.  Perhaps a picture will help illustrate.

Below is an image from my current application.  On the left I have added a menu with 3 options.  When you select one of the options I register/unregister of one of my UiItemsProvider implementations.  When registered, these providers add tab(s) on the right that each contain widgets that display our customer's content in ways useful to them.  The tab circled in red contains two widgets: a ComboBox and a react-table.  The react-table changes its contents based on the currently selected value in the ComboBox widget above it.

For my idea imagine the first menu would have a new option to create a new tab.  This tab would contain 2 widgets: a ComboBox and a ViewPort.  When the user selects a model from the ComboBox (Default3d, sheets, etc.) the viewport is updated to display that model.  Each time the first menu is invoked a new tab would be generated.  (I'd probably add a close button widget do delete the tab when no longer desired.)

Is this a good approach or not?  I don't know.  It's just something I came up with that sort of fits in with the current tab UI pattern.  Where I got stuck was figuring out how to put a ViewPort inside a tab.  _(Given time I might have worked that out but I moved on to higher priority features since I was on a short deadline.)_

![image](https://user-images.githubusercontent.com/30668739/146401386-18e7fb4c-8798-40f7-a622-1325c9bef75c.png)
